psycopg_c/psycopg_c/types/uuid.pyx

```diff
@@ -0,0 +1,138 @@
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 128-143
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 144-159
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 160-175
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 176-191
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 192-207
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 208-223
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  // 224-239
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1   // 240-255
+};
+"""
+    const int8_t[256] hex_to_int_map
+
+
+@cython.final
+cdef class UUIDLoader(CLoader):
+    format = PQ_TEXT
+
+    cdef PyObject *_uuid_type
+    cdef object _uuid_new
+    cdef object _obj_setattr
+    cdef PyObject *_safeuuid_unknown
+
+    def __cinit__(self, oid: int, context: AdaptContext | None = None):
+        global uuid
+        # uuid is slow to import, lazy load it
+        if uuid is None:
+            import uuid
+
+        self._uuid_type = <PyObject *>uuid.UUID
+        self._uuid_new = uuid.UUID.__new__
+        self._obj_setattr = object.__setattr__
+        self._safeuuid_unknown = <PyObject *>uuid.SafeUUID.unknown
+
+    cdef object cload(self, const char *data, size_t length):
+        cdef uint64_t high = 0
+        cdef uint64_t low = 0
+        cdef size_t i
+        cdef int ndigits = 0
+        cdef int8_t c
+
+        for i in range(length):
+            c = data[i]
+            if hex_to_int_map[c] == -1:
+                continue
+
+            if ndigits < 16:
+                high = (high << 4) | hex_to_int_map[c]
+            else:
+                low = (low << 4) | hex_to_int_map[c]
+            ndigits += 1
+
+        if ndigits != 32:
+            raise ValueError("Invalid UUID string")
+
+        cdef object int_value = (PyLong_FromUnsignedLongLong(high) << 64) | PyLong_FromUnsignedLongLong(low)
+
+        cdef object u = PyObject_CallFunctionObjArgs(self._uuid_new, self._uuid_type, NULL)
+        PyObject_CallFunctionObjArgs(self._obj_setattr, <PyObject *>u, <PyObject *>"is_safe", self._safeuuid_unknown, NULL)
+        PyObject_CallFunctionObjArgs(self._obj_setattr, <PyObject *>u, <PyObject *>"int", <PyObject *>int_value, NULL)
+        return u
+
+
+@cython.final
+cdef class UUIDBinaryLoader(CLoader):
+    format = PQ_BINARY
+
+    cdef PyObject *_uuid_type
+    cdef object _uuid_new
+    cdef object _obj_setattr
+    cdef PyObject *_safeuuid_unknown
+
+    def __cinit__(self, oid: int, context: AdaptContext | None = None):
+        global uuid
+        # uuid is slow to import, lazy load it
+        if uuid is None:
+            import uuid
+
+        self._uuid_type = <PyObject *>uuid.UUID
+        self._uuid_new = uuid.UUID.__new__
+        self._obj_setattr = object.__setattr__
+        self._safeuuid_unknown = <PyObject *>uuid.SafeUUID.unknown
```

Comment:
The `__cinit__` is identical for the two loaders. Probably we should introduce a base class from which to derive both.

Probably we can also have an internal method `_uuid_from_int`, so that in the `cload()` we can just deal with parsing the data into the `int_value`, and return self._uuid_from_int(int_value)`.